### PR TITLE
pkp/pkp-lib#3242 Upgrade Smarty to v3

### PR DESCRIPTION
**Pull request by asmecher**

### CODE REVIEW

File: templates/install/install.tpl [hunk 1]
```diff
@@ -10,15 +10,15 @@
...
-{if is_writeable('cache')}{translate|assign:"writable_cache" key="installer.checkYes"}{else}{translate|assign:"writable_cache" key="installer.checkNo"}{/if}
-{if is_writeable('public')}{translate|assign:"writable_public" key="installer.checkYes"}{else}{translate|assign:"writable_public" key="installer.checkNo"}{/if}
-{if is_writeable('cache/_db')}{translate|assign:"writable_db_cache" key="installer.checkYes"}{else}{translate|assign:"writable_db_cache" key="installer.checkNo"}{/if}
-{if is_writeable('cache/t_cache')}{translate|assign:"writable_templates_cache" key="installer.checkYes"}{else}{translate|assign:"writable_templates_cache" key="installer.checkNo"}{/if}
-{if is_writeable('cache/t_compile')}{translate|assign:"writable_templates_compile" key="installer.checkYes"}{else}{translate|assign:"writable_templates_compile" key="installer.checkNo"}{/if}
+{capture assign="writable_config"}{if is_writeable('config.inc.php')}{translate key="installer.checkYes"}{else}{translate key="installer.checkNo"}{/capture}{/if}{/capture}
```
Comment (NateWr) [inline templates/install/install.tpl]: Looks like there may be an extra `{/capture}` here.

File: classes/form/FormBuilderVocabulary.inc.php [hunk 2]
```diff
@@ -750,7 +750,7 @@ function _smartyFBVFileInput($params, &$smarty) {
 	function _smartyFBVKeywordInput($params, &$smarty) {
 		$params['uniqId'] = uniqid();
 
-		$smarty->clear_assign(array('FBV_id', 'FBV_label', 'FBV_label_content', 'FBV_currentKeywords', 'FBV_multilingual', 'FBV_sourceUrl', 'FBV_uniqId', 'FBV_disabled'));
+		$smarty->clearAssign(array('FBV_id', 'FBV_label', 'FBV_label_content', 'FBV_currentKeywords', 'FBV_multilingual', 'FBV_sourceUrl', 'FBV_uniqId', 'FBV_disabled'));
 		$emptyFormLocaleMap = array_fill_keys(array_keys($smarty->get_template_vars('formLocales')), array());
```
Comment (NateWr) [inline classes/form/FormBuilderVocabulary.inc.php]: Looks like there's a `get_template_vars` call that still needs to be updated.

Comment (asmecher) [reply classes/form/FormBuilderVocabulary.inc.php]: Thanks, I fixed that and found a few others.

File: classes/template/PKPTemplateManager.inc.php [hunk 3]
```diff
@@ -61,6 +61,15 @@ class PKPTemplateManager extends Smarty {
...
 	private $_request;
 
+	/** @var string path to the active theme templates */
+	private $theme_template_dir;
+
+	/** @var string path to the old templates */
+	private $app_template_dir;
+
+	/** @var string path to the new templates */
+	private $core_template_dir;
```
[the remark this answers is not in the document]
Comment (asmecher) [reply classes/template/PKPTemplateManager.inc.php]: Corrected!

File: classes/plugins/Plugin.inc.php [hunk 4]
```diff
@@ -312,50 +314,47 @@ function getPluginPath() {
...
 	/**
-	 * Return the Resource Name for templates in this plugin.
-	 *
+	 * Return the Resource Name for templates in this plugin, or if specified, the full resource locator
+	 * for a specific template.
+	 * @param $template Template path/filename, if desired
 	 * @return string
 	 */
-	public function getTemplateResourceName($inCore = false) {
+	public function getTemplateResource($template = null, $inCore = false) {
```
Comment (NateWr) [inline classes/plugins/Plugin.inc.php]: It was already missing, but `$inCore` isn't documented here.

File: classes/plugins/Plugin.inc.php [hunk 5]
```diff
@@ -312,50 +314,47 @@ function getPluginPath() {
...
-	public function _registerTemplateResource($inCore = false) {
-		$templateMgr = TemplateManager::getManager();
-		$pluginPath = $this->getPluginPath();
-		if ($inCore) {
-			$pluginPath = PKP_LIB_PATH . DIRECTORY_SEPARATOR . $pluginPath;
+	protected function _registerTemplateResource($inCore = false) {
+		if ($templatePath = $this->getTemplatePath()) {
+			$templateMgr = TemplateManager::getManager();
+			$pluginTemplateResource = new PKPTemplateResource($templatePath);
+			$templateMgr->registerResource($thing = $this->getTemplateResource(null, $inCore), $pluginTemplateResource);
```
Comment (NateWr) [inline classes/plugins/Plugin.inc.php]: Is `$thing` doing anything here?

File: classes/template/PKPTemplateManager.inc.php [hunk 6]
```diff
@@ -397,16 +382,6 @@ function initialize() {
...
-	/**
-	 * Override the Smarty {include ...} function to allow hooks to be
-	 * called.
-	 */
-	function _smarty_include($params) {
-		if (!HookRegistry::call('TemplateManager::include', array($this, &$params))) {
-			return parent::_smarty_include($params);
-		}
-		return false;
-	}
```
Comment (NateWr) [inline classes/template/PKPTemplateManager.inc.php]: I only found the FidusWriter plugin using this, but maybe just to give him a heads-up?

File: templates/controllers/listbuilder/multipleListsListbuilder.tpl [hunk 7]
```diff
@@ -51,7 +51,7 @@
 					</div>
 				{/if}
 				{assign var=gridTableId value=$staticId|concat:"-table-":$listId}
-				{include file="controllers/listbuilder/listbuilderTable.tpl gridTableId=$gridTableId rows=$listsRows[$listId]}
+				{include file="controllers/listbuilder/listbuilderTable.tpl" gridTableId=$gridTableId rows=$listsRows[$listId]}
```
Comment (NateWr) [inline templates/controllers/listbuilder/multipleListsListbuilder.tpl]: Hah! Nice one.

File: templates/frontend/components/header.tpl [hunk 8]
```diff
@@ -43,11 +43,13 @@
...
-						{else}
-							{url|assign:"homeUrl" context="index" router=$smarty.const.ROUTE_PAGE}
-						{/if}
+						{capture assign="homeUrl"}
+							{if $currentContext && $multipleContexts}
+								{url page="index" router=$smarty.const.ROUTE_PAGE}
+							{else}
+								{url context="index" router=$smarty.const.ROUTE_PAGE}
+							{/if}
+						{/capture}
```
Comment (NateWr) [inline templates/frontend/components/header.tpl]: Is there a scenario where line breaks in a `href` attribute can cause issues? Should `{strip}` be used here? Or can we print `$homeUrl|trim` below maybe?

File: templates/header/usernav.tpl [hunk 9]
```diff
@@ -7,11 +7,13 @@
...
-{elseif $multipleContexts}
-	{url|assign:"homeUrl" context="index" router=$smarty.const.ROUTE_PAGE}
-{/if}
+{capture assign="homeUrl"}
+	{if $currentContext}
+		{url page="index" router=$smarty.const.ROUTE_PAGE}
+	{elseif $multipleContexts}
+		{url context="index" router=$smarty.const.ROUTE_PAGE}
+	{/if}
+{/capture}
```
Comment (NateWr) [inline templates/header/usernav.tpl]: If yes to using `{strip}` or `|trim` above, then do it here too.

File: classes/plugins/Plugin.inc.php (hunk 4 above)
Comment (asmecher) [reply classes/plugins/Plugin.inc.php]: Thanks, fixed.

File: classes/template/PKPTemplateManager.inc.php (hunk 6 above)
Comment (asmecher) [reply classes/template/PKPTemplateManager.inc.php]: Good idea. @johanneswilm, heads-up that the `TemplateManager::include` hook will be changing its conventions with the 3.2 release (late this year).

File: templates/frontend/components/header.tpl (hunk 8 above)
Comment (asmecher) [reply templates/frontend/components/header.tpl]: Smarty seems to remove whitespace when there's just a Smarty tag on the line. I think we're OK as is.

File: classes/template/PKPTemplateManager.inc.php (hunk 6 above)
Comment (johanneswilm) [reply classes/template/PKPTemplateManager.inc.php]: thanks for letting me know!

**Pull request merged**